### ISSUE TITLE
Change how user roles are implemented

### ISSUE THREAD

**Issue by misaugstad** [ProjectSidewalk/SidewalkWebpage#836]
For Administrator and Researcher roles, multiple entries are in the user_role table, one for each role that they have. So if you are just a researcher, you have a User entry and a Researcher entry. If you are an Administrator, you have all 3. I can't think of a scenario wherein this is intuitive and/or useful.

We should consider changing this so that if you become an admin, your role is updated from User to Administrator, instead of adding a new role.

**Comment by jonfroehlich** [comment]
Yes, this should be changed. (Do we know if this change will break anything?).

**Comment by misaugstad** [comment]
Of course there are places where the current implementation dictates how code is written (e.g., using .contains instead of .equals), and these things would have to be rewritten, but I haven't seen anything that would really break if we did this.

**Comment by jonfroehlich** [comment]
Who should do this and when? What's the priority?

**Comment by manaswis** [comment]
This will be done after relaunch. Who will do it has to be decided.

**Comment by misaugstad** [comment]
I think that whoever works on #837 should do this as well, since they will be working with the user_role table anyway.

**Comment by adash12** [comment]
In #837 there was a discussion about adding an "Owner" role for Jon. In that case, the roles would look like this: 
1 - User (no special permissions)
2 - Researcher (has no special permissions but can be excluded from admin charts)
3 - Admin (has special permissions, i.e. accessing admin page and changing user roles)
4 - Owner (has permissions of admin but cannot have role changed by admins)

How does this sound?

**Comment by misaugstad** [comment]
This is what I was thinking of, yep!

**Comment by manaswis** [comment]
@misaugstad 
> Of course there are places where the current implementation dictates how code is written (e.g., using .contains instead of .equals), and these things would have to be rewritten, but I haven't seen anything that would really break if we did this.

Has this been checked? For e.g. admin dashboard -- in functions where we exclude researcher data from the admin dashboard, now "Researcher" data would be from all of these roles "Owner", "Researcher", "Administrator".

**Comment by manaswis** [comment]
E.g. these could be errors due to the changes made:

![image](https://user-images.githubusercontent.com/2873216/31807244-de1ac9f0-b521-11e7-8a0d-2066529a0c99.png)

This shows up on the Analytics tab.

**Comment by manaswis** [comment]
The errors don't show up on the dev server, so should be fine. However, my earlier comments still require a response:

> > @misaugstad: Of course there are places where the current implementation dictates how code is written (e.g., using .contains instead of .equals), and these things would have to be rewritten, but I haven't seen anything that would really break if we did this.
> 
> Has this been checked? For e.g. admin dashboard -- in functions where we exclude researcher data from the admin dashboard, now "Researcher" data would be from all of these roles "Owner", "Researcher", "Administrator".
> 
> Other examples: When a volunteer or a turker signs up, the roles should be appropriately assigned.

**Comment by misaugstad** [comment]
>Has this been checked? For e.g. admin dashboard -- in functions where we exclude researcher data from the admin dashboard, now "Researcher" data would be from all of these roles "Owner", "Researcher", "Administrator".

Oops, I went halfway on this. I made sure that in all those places, we are calling is the function `isResearcher`, which is defined in `UserRoleTable.scala`. I made sure that we were calling this, instead of just getting the role, and checking if it was "Researcher". BUT then I forgot to update the `isResearcher` function to check for all three cases! I will submit a PR momentarily.

>Other examples: When a volunteer or a turker signs up, the roles should be appropriately assigned.

I looked as hard as I could for these things, and all cases *should* have been found. I checked anywhere that any function from the `UserRoleTable.scala` was called, and made sure it was being used correctly. So unless someone decided to add a role to a user by accessing the table directly, from outside the `UserRoleTable.scala` file, we should be covered. And to be clear, that is _highly_ unlikely. The syntax for adding a column to the table is a little weird, and it isn't something that anyone just re-implements elsewhere, especially when we have a strong convention where every `*Table.scala` file has a function to do this.
